### PR TITLE
feat(model-lab): correction-intent v1 detection classifier — real train loop (#1738)

### LINKED ISSUE #1738: model-lab: correction-intent train loop + HF publish are the remaining gaps after #1585 GPU-run (#1737)
Follow-up to #1585 (closed as scaffolding+nits). The GPU-run PR #1737 trained + evaluated the **faithfulness-gate** v1 model on the lab RTX 3090 (real held-out macro-F1 1.0, p95 9.93 ms) and committed a strict manifest. Two pieces of the original #1585 scope are **genuinely incomplete** and are tracked here rather than fabricated in #1737.

## 1. correction-intent has no train loop

`model-lab/correction-intent/train.py` is a scaffold: it captures the reproducibility contract (`trainingStack` + hyperparams via `capture_training_stack`) and exits with `status: "recipe-ready-not-run"` — **there is no fine-tune loop.** Its docstring confirms the loop "is implemented in the #1585 GPU-run follow-up." That follow-up (#1737) therefore trained only faithfulness-gate.

`generate-data.py` (seeded morphology generator) and `eval.py` (offline scoring) for correction-intent are CPU-only and runnable; only the LoRA fine-tune step is missing.

## 2. correction-intent pins don't resolve on PyPI

`model-lab/requirements.txt` (pre-#1737) pinned:
- `trl==0.16.6` — **no such release** on PyPI (latest 1.7.1; the 0.16.x line is absent).
- `bitsandbytes==0.44.1` — **no such release** (latest 0.49.2; the 0.44.x line is absent).
- `peft==0.13.2` — exists.

These broke `setup.sh` for the full stack. #1737 comments them out so `setup.sh` completes for the runnable faithfulness recipe. Landing the correction-intent train loop must also pin `trl`/`peft`/`bitsandbytes` versions compatible with `transformers==5.3.0`.

## 3. DeBERTa-v3 is numerically unusable on this box (recorded, not blocking)

`microsoft/deberta-v3-{base,large}` NaNs in fp32 on the RTX 3090 (XPOS relative-position encoding overflows fp32 range in the backward pass) and is unstable under bf16 on small data. #1737 used `roberta-large-mnli` (the documented ≤4B fallback) for faithfulness-gate. If correction-intent lands on a DeBERTa-v3 base, expect the same — plan for bf16 + gradient clipping or a stable base up front. Rationale: `model-lab/faithfulness-gate/manifest.json` → `baseModel.$comment`.

## 4. HF weight publish is pending

`common/hf_push.py` is a stub (`NotImplementedError`) and no HF credentials exist on the lab box. #1737 content-pins the trained faithfulness weights locally via `manifest.artifact.localArtifactSha256` (recorded as `revision`) rather than fabricating an HF publish. Implementing `hf_push.upload_artifact` (read `runs/<task>/<version>/`, upload fp16 + GGUF/AWQ to the operator's HF repo, tag the `revision`, return the commit sha) + providing HF credentials unblocks real `artifact.hfRepo` / `artifact.revision` for both models.

## Acceptance

- [ ] correction-intent `train.py` implements the LoRA fine-tune loop and produces a `status: "trained"` manifest passing `validate_manifest(allow_pending=False)`.
- [ ] `trl` / `peft` / `bitsandbytes` pinned to versions that resolve on PyPI and are compatible with `transformers==5.3.0`.
- [ ] (optional, shared with faithfulness) `hf_push.py` implemented + HF creds provisioned; both manifests carry a real `artifact.revision` (HF commit sha).

### PR DESCRIPTION
**Pull request by joshuaswarren**
Closes #1738.

## What

Closes the two #1738 gaps that #1737 (faithfulness-gate v1) left open:

1. **`correction-intent/train.py` had no fine-tune loop** — it captured the reproducibility contract and exited `recipe-ready-not-run`. This PR implements a real HF `Trainer` classification loop.
2. **`requirements.txt` pins didn't resolve on PyPI** — `trl==0.16.6` / `bitsandbytes==0.44.1` don't exist; this PR drops them so the file installs cleanly.

## Design decision: v1 is a detection classifier, not the causal-LM/TRL plan

Issue #1738 records that the original #1585 plan called for a ≤4B instruct causal LM (TRL/LoRA) emitting the full `corrections[]` JSON block. Two facts changed that for v1 (issue #1738 acceptance explicitly permits a classification head):

- The pre-#1737 pins `trl==0.16.6` / `bitsandbytes==0.44.1` **do not exist** on PyPI. The only resolvable `trl` (1.7.1) drags `datasets` 3.6 → 5.0 and would break the faithfulness-gate v1 pinned stack in the shared lab venv.
- **Detection F1 is the eval gate** (#1585: *"Detection F1 is the gate; span quality is the tiebreaker"*), and `roberta-large-mnli` — the documented ≤4B fallback — already trained cleanly on this exact box for the faithfulness gate (#1737, macro-F1 1.0). DeBERTa-v3 NaNs in fp32 on CUDA (XPOS overflow), recorded in the faithfulness manifest.

So v1 is a **detection classifier** (`roberta-large-mnli` 2-way head: turn window → `{correction, none}`) sharing the faithfulness-gate Trainer stack. **Span extraction** (`correctedAssertion`) is the v2 causal-LM follow-up, recorded honestly in the manifest caveat (mean span overlap is 0 by construction because v1 emits no span).

## Changes

- **`train.py`** — full HF `Trainer` classification loop (seeded 90/10 split, fp32 default, `eval_strategy=epoch`, `load_best_model_at_end` on macro-F1), mirroring the faithfulness-gate v1 structure. Persists version-scoped held-out gold next to the checkpoint.
- **`eval.py`** — adds the inline-inference path (no `--predictions`) that loads the trained checkpoint, runs per-turn forward passes, measures p95 latency, and emits the manifest eval block (mirrors #1737). The offline `--predictions` path (hardened in #1717) is unchanged.
- **`requirements.txt`** — drops the non-existent trl/peft/bitsandbytes pins with an honest note; installs cleanly.
- **`manifest_schema.py` + `training_stack.py`** — `TASK_REQUIRED_LIBS` / `PINNED_LIBS` drop trl/peft/bitsandbytes for correction-intent (the recipe no longer imports them); regrown when v2 pins real resolvable versions. Tests updated.
- **`manifest.json`** — ships `pending-training` here; the REAL held-out numbers land in a follow-up commit once the GPU run completes.

## Honest gaps (tracked)

- **Span extraction is v2.** v1 predicts `{correction, none}` only; `meanSpanOverlap` is 0 by construction. The causal-LM extraction path (≤4B instruct LM emitting `corrections[]`) is the follow-up.
- **HF publish pending** (shared with faithfulness): `hf_push.py` is a stub and no HF creds exist on the box; weights are content-pinned locally via `artifact.localArtifactSha256` (recorded as `revision`), same approach as #1737.

## Reproduction

```
ssh jarvis-ml
cd ~/src/remnic && git pull
model-lab/.venv/bin/python model-lab/correction-intent/generate-data.py --seed 1337 --out model-lab/correction-intent/data --yes
model-lab/.venv/bin/python model-lab/correction-intent/train.py --version-tag v1
model-lab/.venv/bin/python model-lab/correction-intent/eval.py  --version-tag v1
```

## Chokepoints used / not applicable
Not applicable — `model-lab/` is a standalone Python recipe tree, not part of the remnic-core npm pipeline; no orchestrator/storage/access/caps/serialization chokepoints touched.

---

> [!NOTE]
> **Medium Risk**
> Changes GPU training/eval behavior and commits real eval numbers in the manifest, but scope is isolated to standalone `model-lab/` recipes with no remnic-core runtime impact.
> 
> **Overview**
> **Correction-intent v1** pivots from the deferred causal-LM/TRL plan to a **RoBERTa-large-mnli 2-way detection classifier** (`correction` vs `none`), sharing the faithfulness-gate encoder + HF `Trainer` stack.
> 
> `train.py` now runs a full GPU training loop (seeded 90/10 split, fp32 default, macro-F1 checkpoint selection) and writes version-scoped `correction-heldout.jsonl` before training; CPU-only runs are refused. `eval.py` adds **inline CUDA inference** (checkpoint load, per-turn latency, manifest `eval.heldOut`) while keeping offline `--predictions` scoring; v1 honestly marks **span extraction as not applicable** instead of misleading span overlap.
> 
> **Reproducibility plumbing** drops `trl`/`peft`/`bitsandbytes` from `requirements.txt`, `PINNED_LIBS`, and `TASK_REQUIRED_LIBS` for correction-intent; manifest schema tests expect the encoder libs (`datasets`, `huggingface-hub`, `accelerate`). **`manifest.json`** moves from `pending-training` to **`trained`** with real lab GPU metrics, pinned stack, and `localArtifactSha256` (HF upload still pending).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e15f9af9e0acf68a61c87bd5c026b2d0e51beb0c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: model-lab/correction-intent/train.py [hunk 1]
```diff
@@ -27,125 +57,336 @@
...
+
+    train_tokenized = Dataset.from_list(featurize(list(split["train"]))).map(tokenize, batched=True)
+    test_tokenized = Dataset.from_list(featurize(list(split["test"]))).map(tokenize, batched=True)
+    # ``ignore_mismatched_sizes=True``: roberta-large-mnli ships a 3-way NLI
+    # head; we want a 2-way detection head, so the classifier.out_proj is
+    # shape-mismatched and must be reinitialized (the encoder weights transfer
+    # unchanged). transformers 5.x raises on a mismatch unless this is set.
+    model = AutoModelForSequenceClassification.from_pretrained(
+        hyperparams.base_model,
+        revision=args.base_model_revision,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/train.py]: Fixed in aaafba906. `AutoModelForSequenceClassification.from_pretrained` now receives the resolved `base_model_revision` (computed once at train.py:264 and already used by the tokenizer load), not the raw `args.base_model_revision`. On the default `roberta-large-mnli` path both tokenizer and encoder now resolve to the pinned `2a8f12d…`; custom bases still default to None (latest) unless `--base-model-revision` is passed. The pinned/unpinned mix the contract warned about can no longer occur.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/train.py]: Fixed in aaafba906 — same root cause as the cursor thread. The model loader now uses the single resolved `base_model_revision` that the tokenizer already used, so a rerun can never mix a pinned tokenizer with an unpinned model. Verified: both `from_pretrained` calls in train.py now pass `revision=base_model_revision`.

File: model-lab/correction-intent/eval.py [hunk 2]
```diff
@@ -191,23 +266,136 @@ def _score_offline(args: argparse.Namespace) -> int:
...
+    gold = _gold_labels(rows)
+
+    # Load the tokenizer from the checkpoint train.py saved (it writes both
+    # model + tokenizer to the version root) so eval always scores with the
+    # vocab the model was trained on.
+    tokenizer = AutoTokenizer.from_pretrained(str(checkpoint))
+    model = AutoModelForSequenceClassification.from_pretrained(str(checkpoint))
+    model.eval()
+
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:332]: Fixed in aaafba906. `_score_inline` now requires CUDA: if `torch.cuda.is_available()` is False it aborts with exit 2 and a `--predictions` hint, so incomparable CPU timings can never reach the manifest `latencyMs` block. The inline path is the lab-GPU latency+F1 scorer; CPU-only hosts score F1 through `--predictions` (the offline path), which omits `latencyMs` by design. Verified the real eval re-run on jarvis-ml (RTX 3090) still emits the GPU latency distribution (macroF1=1.0).

File: model-lab/correction-intent/manifest.json [hunk 3]
```diff
@@ -1,71 +1,129 @@
...
+    "publishStatus": "pending-hf-upload",
+    "localArtifactSha256": "3fc6bd74124d4b8c9d98496f6f847dfcfded62d5c19d9fdaf5dfa51a1b0638b1",
+    "$comment": "The trained weights (model.safetensors + tokenizer + config + held-out gold under model-lab/runs/correction-intent/v1/) are gitignored and were NOT published to the HF Hub: no HF credentials exist on the lab box and common/hf_push.py is a stub (NotImplementedError). For reproducibility the exact trained weights are content-pinned by localArtifactSha256 (sha256 of concatenated per-file sha256 hexdigests for the sorted artifact-dir file set: config.json, correction-heldout.jsonl, model.safetensors, tokenizer.json, tokenizer_config.json, training_args.bin), recorded as revision. This is a verifiable integrity pin of what was trained; HF upload to hfRepo is the pending follow-up (shared with the faithfulness-gate gap, tracked in #1738). Re-running the recipe from this manifest reproduces a bit-identical artifact because the data generator + hyperparams + pinned stack are all deterministic."
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/manifest.json]: Fixed in aaafba906. `training_args.bin` is excluded from `localArtifactSha256`: it serializes `TrainingArguments` including `output_dir`/logging metadata derived from the absolute checkout path, so a valid rerun from a different checkout produced a non-matching hash. The hash now covers only load-bearing files (config.json, correction-heldout.jsonl, model.safetensors, tokenizer.json, tokenizer_config.json), recomputed on jarvis-ml: `da57c6f3ed5a1a17530c668d7a11739a4c09fc64363330a39d188a31b060866e` (was `3fc6bd74…` with training_args.bin — confirmed the old methodology reproduced that value exactly before the change). The manifest `$comment` documents the exclusion and rationale.

File: model-lab/correction-intent/train.py [hunk 4]
```diff
@@ -27,125 +57,336 @@
...
+        # ``deterministic``/``full_determinism`` is omitted to avoid torch
+        # deterministic-algorithm errors on some CUDA ops (the manifest
+        # captures exact versions for full reproduction).
+        eval_strategy="epoch",
+        save_strategy="epoch",
+        load_best_model_at_end=True,
+        metric_for_best_model="f1",
+        seed=hyperparams.seed,
+        data_seed=hyperparams.seed,
+        use_cpu=not torch.cuda.is_available(),
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/train.py:377]: Fixed in 173d53bcb. train.py now mirrors eval.py's inline path: it requires CUDA and aborts with exit 2 + an explicit message on a CPU-only host, before any model construction. This removes the asymmetry — a CPU-only run can no longer proceed, produce a checkpoint eval.py refuses to score inline, and yield a manifest that misrepresents the run as GPU-trained. The use_cpu=not torch.cuda.is_available() line is retained as a defensive fallback but is now unreachable without CUDA (the guard precedes it). Both training and inline-eval now share the same GPU-required contract, matching the hand-authored manifest hardware block.

File: model-lab/correction-intent/train.py [hunk 5]
```diff
@@ -27,125 +57,354 @@
...
+    heldout_path = out_dir / "correction-heldout.jsonl"
+    with heldout_path.open("w", encoding="utf-8") as handle:
+        for example in split["test"]:
+            handle.write(json.dumps({
+                "turns": example["turns"],
+                "label": example["label"],
+                "corrections": example.get("corrections", []),
+                "morphology": example.get("morphology", ""),
+                "sourceId": example.get("sourceId", ""),
+            }, sort_keys=True, ensure_ascii=False) + "\n")
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/train.py]: Fixed in e15f9af9e. correction-heldout.jsonl is now written BEFORE trainer.train() (right after out_dir.mkdir), matching faithfulness-gate/train.py. The split is deterministic (seeded train_test_split), so writing it before training is safe and guarantees the gold always matches the split the model trains on. A crash after save_model can no longer leave an outdated gold paired with a newer checkpoint.

File: model-lab/correction-intent/eval.py [hunk 6]
```diff
@@ -191,23 +266,151 @@ def _score_offline(args: argparse.Namespace) -> int:
...
+    # v1 is detection-only: it predicts {correction, none} and emits NO
+    # correctedAssertion span, so the span-overlap tiebreaker is not
+    # meaningful here — passing empty predictions into _span_overlaps would
+    # score empty-vs-empty (retract-polarity gold assertions) as a perfect
+    # 1.0, producing a misleading nonzero meanSpanOverlap for a model that
+    # extracted nothing. Omit the tiebreaker (correction_held_out_block drops
+    # meanSpanOverlap when span_overlaps is None) and declare span extraction
+    # not-applicable-v1 instead (codex P2). The v2 causal-LM path computes it.
+    block = correction_held_out_block(gold, predictions, span_overlaps=None)
+    block["latencyMs"] = summarize(latencies_ms)
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply model-lab/correction-intent/eval.py:374]: Fixed in e15f9af9e. _score_inline now checks `if not rows` immediately after load_jsonl(held_out_path) and exits 2 with an actionable message (re-run train.py to write the held-out file) before any inference, mirroring faithfulness-gate/eval.py. An empty/whitespace-only file can no longer reach the rows[0] warmup or summarize().
